src/webgpu/shader/validation/expression/call/builtin/select.spec.ts

```diff
@@ -0,0 +1,232 @@
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import { validateConstOrOverrideBuiltinEval } from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kArgumentTypes = objectsToRecord(kAllScalarsAndVectors);
+
+g.test('argument_types_1_and_2')
+  .desc(
+    `
+Validates that scalar and vector arguments are not rejected by ${builtin}() for args 1 and 2
+`
+  )
+  .params(u => u.combine('type1', keysOf(kArgumentTypes)).combine('type2', keysOf(kArgumentTypes)))
+  .beforeAllSubcases(t => {
+    if (
+      scalarTypeOf(kArgumentTypes[t.params.type1]) === Type.f16 ||
+      scalarTypeOf(kArgumentTypes[t.params.type2]) === Type.f16
+    ) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type1 = kArgumentTypes[t.params.type1];
+    const type2 = kArgumentTypes[t.params.type2];
+    // First and second arg must be the same or one convertible to the other.
+    // Note that we specify a concrete return type even if both args are abstract.
+    const returnType = isConvertible(type1, type2)
+      ? concreteTypeOf(type2)
+      : isConvertible(type2, type1)
+      ? concreteTypeOf(type1)
+      : undefined;
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      /* expectedResult */ returnType !== undefined,
+      [type1.create(0), type2.create(0), Type.bool.create(0)],
+      'constant',
+      returnType
+    );
+  });
+
+g.test('argument_types_3')
+  .desc(
+    `
+Validates that third argument must be bool for ${builtin}()
+`
+  )
+  .params(u => u.combine('type', keysOf(kArgumentTypes)))
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    const type = kArgumentTypes[t.params.type];
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      /* expectedResult */ type === Type.bool,
+      [Type.i32.create(0), Type.i32.create(0), type.create(0)],
+      'constant',
+      /*return_type*/ Type.i32
+    );
+  });
+
+const kTests = {
+  valid: {
+    src: `_ = ${builtin}(1, 2, true);`,
+    pass: true,
+  },
+  alias: {
+    src: `_ = ${builtin}(i32_alias(1), i32_alias(2), bool_alias(true));`,
+    pass: true,
+  },
+  bool: {
+    src: `_ = ${builtin}(false, false, true);`,
+    pass: true,
+  },
+  i32: {
+    src: `_ = ${builtin}(1i, 1i, true);`,
+    pass: true,
+  },
+  u32: {
+    src: `_ = ${builtin}(1u, 1u, true);`,
+    pass: true,
+  },
+  f32: {
+    src: `_ = ${builtin}(1.0f, 1.0f, true);`,
+    pass: true,
+  },
+  f16: {
+    src: `_ = ${builtin}(1.0h, 1.0h, true);`,
+    pass: true,
+  },
+  mixed_aint_afloat: {
+    src: `_ = ${builtin}(1, 1.0, true);`,
+    pass: true,
+  },
+  mixed_i32_u32: {
+    src: `_ = ${builtin}(1i, 1u, true);`,
+    pass: false,
+  },
+  vec_bool: {
+    src: `_ = ${builtin}(vec2<bool>(false, true), vec2<bool>(false, true), true);`,
```

Comment:
Partially covered with your other test, but it would also be nice to have a few inferred type `vecN()` arguments in here - especially when the element type has to be inferred from another other argument.

Reply:
Done.